packages/data-flow/src/retroactiveProcessor.ts

```diff
@@ -0,0 +1,320 @@
+        );
+    }
+
+    /**
+     * Process historical events for a specific strategy
+     * @param strategyId - Identifier of the strategy to process
+     * @param strategyAddresses - Set of contract addresses implementing this strategy
+     * @param lastEventPointer - Latest processed event pointer to process up to
+     */
+    private async processRetroactiveStrategy(
+        strategyId: Hex,
+        strategyAddresses: Set<Address>,
+        lastEventPointer: Readonly<EventPointer>,
+    ): Promise<void> {
+        // Check if we have a checkpoint for this strategy
+        const checkpoint = await this.checkpointRepository.getCheckpoint(this.chainId, strategyId);
+
+        const currentPointer: EventPointer = checkpoint
+            ? {
+                  blockNumber: checkpoint.lastProcessedBlockNumber,
+                  logIndex: checkpoint.lastProcessedLogIndex,
+              }
+            : { blockNumber: 0, logIndex: 0 };
+
+        const events = new Queue<ProcessorEvent<ContractName, AnyEvent> & { strategyId?: Hex }>();
+        let event: (ProcessorEvent<ContractName, AnyEvent> & { strategyId?: Hex }) | undefined;
+
+        while (true) {
+            try {
+                await this.enqueueEventsIfEmpty(
+                    events,
+                    strategyAddresses,
+                    currentPointer,
+                    lastEventPointer,
+                );
+
+                event = events.pop();
+                if (!event) break;
+
+                currentPointer.blockNumber = event.blockNumber;
+                currentPointer.logIndex = event.logIndex;
+
+                if (this.hasReachedLastEvent(currentPointer, lastEventPointer)) break;
+
+                event.strategyId = strategyId;
+                const changesets = await this.eventsProcessor.processEvent(event);
+                const executionResult = await this.dataLoader.applyChanges(changesets);
+
+                if (executionResult.numFailed > 0) {
+                    this.logger.error(
+                        `Failed to apply changesets. ${executionResult.errors.join("\n")} Event: ${stringify(event)}`,
+                    );
+                }
+            } catch (error) {
+                if (error instanceof InvalidEvent || error instanceof UnsupportedEventException) {
+                    // Expected errors that we can safely ignore
+                    this.logger.debug(`Skipping error for ${error.name}: ${stringify(event)}`);
+                } else {
+                    this.logger.error(`Error processing event: ${stringify(event)} ${error}`);
+                }
+            }
+
+            // Update checkpoint after processing
+            await this.updateCheckpoint(strategyId, currentPointer);
+        }
+
+        await this.markStrategyAsHandled(strategyId, strategyAddresses);
+        // Delete checkpoint after processing of all events
+        await this.checkpointRepository.deleteCheckpoint(this.chainId, strategyId);
```

Comment:
I'm sure I'm getting this wrong due to not being familiar with the codebase but wouldn't `deleteCheckpoint` delete the updated checkpoint on line 214? Or the idea is to use these checkpoints just for this retroactive processing?

Reply:
as you're saying, the checkpoints are specific for this retroactive processing in case smth critical happens (like machine goes down), so we restart and keep processing from the point we were. this won't be used a lot actually and when used is not that we will have tons of new strategies